src/queryHistory/queryHistoryProvider.ts

```diff
@@ -0,0 +1,206 @@
+        private _outputContentProvider: SqlOutputContentProvider,
+        private _vscodeWrapper: VscodeWrapper,
+        private _untitledSqlDocumentService: UntitledSqlDocumentService,
+        private _statusView: StatusView,
+        private _prompter: IPrompter
+    ) {
+        const config = this._vscodeWrapper.getConfiguration(Constants.extensionConfigSectionName);
+        this._queryHistoryLimit = config.get(Constants.configQueryHistoryLimit);
+        this._queryHistoryEnabled = config.get(Constants.configEnableQueryHistoryCapture);
+        this._vscodeWrapper.setContext(Constants.isQueryHistoryEnabled, this._queryHistoryEnabled);
+        this._queryHistoryUI = new QueryHistoryUI(this._prompter, this._vscodeWrapper);
+    }
+
+    clearAll(): void {
+        this._queryHistoryNodes = [new EmptyHistoryNode()];
+        this._onDidChangeTreeData.fire();
+    }
+
+    refresh(ownerUri: string, timeStamp: Date, hasError): void {
+        const timeStampString = timeStamp.toLocaleString();
+        const historyNodeLabel = this.createHistoryNodeLabel(ownerUri, timeStampString);
+        const tooltip = this.createHistoryNodeTooltip(ownerUri, timeStampString);
+        const queryString = this.getQueryString(ownerUri);
+        const connectionLabel = this.getConnectionLabel(ownerUri);
+        const node = new QueryHistoryNode(historyNodeLabel, tooltip, queryString,
+            ownerUri, timeStamp, connectionLabel, !hasError);
+        if (this._queryHistoryNodes.length === 1) {
+            if (this._queryHistoryNodes[0] instanceof EmptyHistoryNode) {
+                this._queryHistoryNodes = [];
+            }
+        }
+        this._queryHistoryNodes.push(node);
+        // Push out the first listing if it crosses limit to maintain
+        // an LRU order
+        if (this._queryHistoryNodes.length > this._queryHistoryLimit) {
+            this._queryHistoryNodes.shift();
+        }
+        // return the query history sorted by timestamp
+        this._queryHistoryNodes.sort((a, b) => {
+            return (b as QueryHistoryNode).timeStamp.getTime()-
+                (a as QueryHistoryNode).timeStamp.getTime();
+        });
+        this._onDidChangeTreeData.fire();
+    }
+
+    getTreeItem(node: QueryHistoryNode): any {
+        return node;
+    }
+
+    async getChildren(element?: any): Promise<vscode.TreeItem[]> {
+        if (this._queryHistoryNodes.length == 0) {
+            this._queryHistoryNodes.push(new EmptyHistoryNode());
+        }
+        return this._queryHistoryNodes;
+    }
+
+    /**
+     *
+     */
+    public async showQueryHistoryCommandPalette(): Promise<void> {
+        const options = this._queryHistoryNodes.map(node => this._queryHistoryUI.convertToQuickPickItem(node));
+        let queryHistoryQuickPickItem = await this._queryHistoryUI.showQueryHistoryCommandPalette(options);
+        this.openQueryHistoryEntry(queryHistoryQuickPickItem.node, queryHistoryQuickPickItem.action ===
```

Comment:
If the user cancels out of the dialog is queryHistoryPickItem going to be undefined?

Reply:
There will be a rejected promise in that case. I'll add a check just to be sure.